### PR TITLE
[Don't merge] fix dynamic size array(vla) build error on msvc v2

### PR DESCRIPTION
**Pull request by xuhancn**
MSVC don't support dynamic array.
Ref: https://stackoverflow.com/questions/56555406/creating-dynamic-sized-array-using-msvc-c-compiler 

We tried to use std::vector to instead of it in previous PR: https://github.com/pytorch/pytorch/pull/134140, but it changed variable's type and failed at UTs.

So we had to backup solution, use `std::unique_ptr` to instead of VLA. 

Reproduce UT:
```cmd
pytest test/inductor/test_cpu_repro.py -v -k test_reduction_with_dynamic_threads
```

Error msg:
```cmd
C:/Users/Xuhan/AppData/Local/Temp/tmpncykej5v/a4/ca4534cazplidnf7vopaaxaifqkjiyhxm3h2gsylgztputbaeybx.cpp(13): error C2131: expression did not evaluate to a constant
C:/Users/Xuhan/AppData/Local/Temp/tmpncykej5v/a4/ca4534cazplidnf7vopaaxaifqkjiyhxm3h2gsylgztputbaeybx.cpp(13): note: failure was caused by a read of a variable outside its lifetime
C:/Users/Xuhan/AppData/Local/Temp/tmpncykej5v/a4/ca4534cazplidnf7vopaaxaifqkjiyhxm3h2gsylgztputbaeybx.cpp(13): note: see usage of 'max_threads'
C:/Users/Xuhan/AppData/Local/Temp/tmpncykej5v/a4/ca4534cazplidnf7vopaaxaifqkjiyhxm3h2gsylgztputbaeybx.cpp(16): error C3863: array type 'float [max_threads]' is not assignable
```
Genarated code:
```c++

#include "C:/Users/Xuhan/AppData/Local/Temp/tmpt6mxcjzi/j2/cj22tgrdgh42wbunl7gdptg2lintcziox2kmr7rdbcc6n2njrhgx.h"
extern "C" __declspec(dllexport) void kernel(const float* in_ptr0,
                       const float* in_ptr1,
                       float* out_ptr0,
                       float* out_ptr1)
{
    {
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            int max_threads = omp_get_max_threads();
            float tmp_acc0_arr[max_threads];
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_arr[tid] = 0;
            }
            at::vec::Vectorized<float> tmp_acc0_vec_arr[max_threads];
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_vec_arr[tid] = at::vec::Vectorized<float>(0);
            }
            #pragma omp parallel
```

Fixed by `std::unique_ptr`:
```c++

#include "C:/Users/Xuhan/AppData/Local/Temp/tmphhoo8k5c/j2/cj22tgrdgh42wbunl7gdptg2lintcziox2kmr7rdbcc6n2njrhgx.h"
extern "C" __declspec(dllexport) void kernel(const float* in_ptr0,
                       const float* in_ptr1,
                       float* out_ptr0,
                       float* out_ptr1)
{
    {
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            int max_threads = omp_get_max_threads();
            std::unique_ptr<float[]> tmp_acc0_arr(new float[max_threads]);
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_arr[tid] = 0;
            }
            std::unique_ptr<at::vec::Vectorized<float>[]> tmp_acc0_vec_arr(new at::vec::Vectorized<float>[max_threads]);
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_vec_arr[tid] = at::vec::Vectorized<float>(0);
            }
            #pragma omp parallel
```

### Local test:
1. `pytest test/inductor/test_cpu_repro.py -v -k test_reduction_with_dynamic_threads` :
image

2. https://github.com/pytorch/pytorch/pull/134140 failed UT: `pytest test/inductor/test_cpu_repro.py  -v -k test_bool_reduction_vec`:
image

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1570,12 +1570,18 @@ def _gen_parallel_reduction_buffers(
...
             "max_threads" if config.cpp.dynamic_threads else parallel_num_threads()
         )
-        acc_per_thread = f"{acc}_arr[{num_threads}]"
+        acc_per_thread_var_name = f"{acc}_arr"
+        acc_per_thread = f"{acc_per_thread_var_name}[{num_threads}]"
+        """
+        MSVC don't support dynamic array(VLA). Please use std::unique_ptr to instead of it.
+        Ref: https://stackoverflow.com/questions/56555406/creating-dynamic-sized-array-using-msvc-c-compiler
+        """
+        acc_per_thread_decl = f"std::unique_ptr<{acc_type}[]> {acc_per_thread_var_name}(new {acc_type}[{num_threads}])"
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1579]:
```suggestion
        acc_per_thread_decl = f"auto {acc_per_thread_var_name} = std::make_unique<{acc_type}[]>({num_threads})"
```

**Review by jgong5 — APPROVED** [review]
A minor suggestion to simplify things. Others LGTM.

**Review by jansel — CHANGES_REQUESTED** [review]
Can we do this without a heap allocation (call to new/malloc)?

I'd expect this version to be slower than the prior one since you added a dynamic memory allocation.

I'd expect the number of max number of threads to be a small constant, we could just over-allocate to the max supported threads.

**Comment by jgong5** [comment]
> Can we do this without a heap allocation (call to new/malloc)?
> 
> I'd expect this version to be slower than the prior one since you added a dynamic memory allocation.
> 
> I'd expect the number of max number of threads to be a small constant, we could just over-allocate to the max supported threads.

Perhaps we can use a fixed-sized array by default and fall back to a dynamically allocated array if the pre-defined size is not enough?

**Pull request closed without merging**